### PR TITLE
[SYCL] Add tests for user-defined reductions extension

### PR DESCRIPTION
**Pull request by dm-vodopyanov**
Spec: https://github.com/intel/llvm/pull/7202
Implementation: https://github.com/intel/llvm/pull/7587

### CODE REVIEW

File: UserDefinedReductions/user_defined_reductions.cpp [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+      accessor outputValues{outputBuf, h, write_only, no_init};
+
+      constexpr size_t group_size = 16;
+
+      // Create enough local memory for the algorithm
+      constexpr size_t temp_memory_size = group_size * sizeof(int);
+      auto scratch = sycl::local_accessor<std::byte, 1>(temp_memory_size, h);
+
+      h.parallel_for(
+          sycl::nd_range<1>(range<1>(16), range<1>(16)), [=](nd_item<1> it) {
```
Comment (aelovikov-intel) [inline UserDefinedReductions/user_defined_reductions.cpp]: Should it use `group_size` instead of `16`?

File: UserDefinedReductions/user_defined_reductions.cpp [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+                    UserDefinedSum<int>{});
+            outputValues[1] = sum_reduce_over_group;
+          });
+    });
+    q.wait();
+  }
+  assert(outputBuf.get_host_access()[0] == 523776);
+  assert(outputBuf.get_host_access()[1] == 120);
+
+  // TODO: add tests for reduce_over_group overloads ith native binary_op
```
Comment (aelovikov-intel) [inline UserDefinedReductions/user_defined_reductions.cpp]:
```suggestion
  // TODO: add tests for reduce_over_group overloads with native binary_op
```
maybe?

File: UserDefinedReductions/user_defined_reductions.cpp [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+            int sum_reduce_over_group =
+                sycl::ext::oneapi::experimental::reduce_over_group(
+                    handle, inputValues[it.get_global_id(0)], 0,
+                    UserDefinedSum<int>{});
+            outputValues[1] = sum_reduce_over_group;
+          });
+    });
+    q.wait();
+  }
+  assert(outputBuf.get_host_access()[0] == 523776);
```
Comment (aelovikov-intel) [inline UserDefinedReductions/user_defined_reductions.cpp]: I'd rather write it as "(min + max) * N /2" or whatever the formula is.

File: UserDefinedReductions/user_defined_reductions.cpp [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+      h.parallel_for(
+          sycl::nd_range<1>(range<1>(16), range<1>(16)), [=](nd_item<1> it) {
+            // Create a handle that associates the group with an allocation it
+            // can use
+            auto handle =
+                sycl::ext::oneapi::experimental::group_with_scratchpad(
+                    it.get_group(), sycl::span(&scratch[0], temp_memory_size));
+
+            int *first = inputValues.get_pointer();
+            int *last = first + 1024;
```
Comment (aelovikov-intel) [inline UserDefinedReductions/user_defined_reductions.cpp]: Need more tests with not so nice numbers/WG sizes/etc.

File: UserDefinedReductions/user_defined_reductions.cpp [hunk 5]
```diff
@@ -0,0 +1,68 @@
+// RUN: %clangxx -fsycl %s -o %t.out
```
Comment (aelovikov-intel) [inline UserDefinedReductions/user_defined_reductions.cpp]: Should be under SYCL/ director, I believe.

Comment (dm-vodopyanov) [reply UserDefinedReductions/user_defined_reductions.cpp]: Thanks for finding, accidentally created a folder in wrong directory.

File: SYCL/UserDefinedReductions/user_defined_reductions.cpp [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+// RUN: %t.out
+
+#include <numeric>
+
+#include <sycl/ext/oneapi/experimental/user_defined_reductions.hpp>
+#include <sycl/sycl.hpp>
+
+template <typename T> struct UserDefinedSum {
+  T operator()(T a, T b) { return a + b; }
+};
```
Comment (aelovikov-intel) [inline SYCL/UserDefinedReductions/user_defined_reductions.cpp:11]: I think we need to test different op/identity/default-init value (e.g., unequal to identity/deleted default ctor if allowed).

Comment (dm-vodopyanov) [reply SYCL/UserDefinedReductions/user_defined_reductions.cpp:11]: Added in https://github.com/intel/llvm-test-suite/pull/1395/commits/760338ccf459ff8a64aae77a94ded0fc7c30cb7c

File: SYCL/UserDefinedReductions/user_defined_reductions.cpp [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+using namespace sycl;
+
+template <typename InputContainer, typename OutputContainer,
+          class BinaryOperation>
+void test(queue q, InputContainer input, OutputContainer output,
+          BinaryOperation binary_op,
+          typename OutputContainer::value_type identity,
+          typename OutputContainer::value_type init) {
+  typedef typename InputContainer::value_type InputT;
+  typedef typename OutputContainer::value_type OutputT;
```
Comment (aelovikov-intel) [inline SYCL/UserDefinedReductions/user_defined_reductions.cpp]: Prefer `using AliasTy = <...>;`

File: SYCL/UserDefinedReductions/user_defined_reductions.cpp [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+}
+
+int main() {
+  queue q;
+
+  constexpr int N = 128;
+  std::array<int, N> input;
+  std::array<int, 4> output;
+  std::iota(input.begin(), input.end(), 0);
+  std::fill(output.begin(), output.end(), 0);
```
Comment (aelovikov-intel) [inline SYCL/UserDefinedReductions/user_defined_reductions.cpp:160]: Why are those created outside `test`?

File: UserDefinedReductions/user_defined_reductions.cpp (hunk 4 above)
Comment (dm-vodopyanov) [reply UserDefinedReductions/user_defined_reductions.cpp]: Applied in https://github.com/intel/llvm-test-suite/pull/1395/commits/597687b80c40d73cd75223c9c800ef96ff841e32

File: SYCL/UserDefinedReductions/user_defined_reductions.cpp (hunk 7 above)
Comment (dm-vodopyanov) [reply SYCL/UserDefinedReductions/user_defined_reductions.cpp]: Applied in https://github.com/intel/llvm-test-suite/pull/1395/commits/597687b80c40d73cd75223c9c800ef96ff841e32

File: SYCL/UserDefinedReductions/user_defined_reductions.cpp (hunk 8 above)
Comment (dm-vodopyanov) [reply SYCL/UserDefinedReductions/user_defined_reductions.cpp:160]: The preparation of data is not part of `test` func because data initialization is different for different data structures: usual arrays of ints, arrays of custom data structures with and without default constructor, arrays with complex numbers

File: SYCL/UserDefinedReductions/user_defined_reductions.cpp [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+  assert(output[0] == std::reduce(input.begin(), input.begin() + workgroup_size,
+                                  identity, binary_op));
+  assert(output[1] == std::reduce(input.begin(), input.begin() + workgroup_size,
+                                  init, binary_op));
+  assert(output[2] ==
+         std::reduce(input.begin(), input.end(), identity, binary_op));
+  assert(output[3] == std::reduce(input.begin(), input.end(), init, binary_op));
+}
+
+int main() {
```
Comment (aelovikov-intel) [inline SYCL/UserDefinedReductions/user_defined_reductions.cpp:153]: Please add a test with reduce over sub_group when one (or even several) WG(s) contain(s) several SGs. Might be done in another PR. The main thing I want is to use such a test as an example showing that even for reduce over SG the size of the temporary memory depends on WG size (unless we can allocate private memory for the scratchpad somehow and then use sg shuffles in the implementation).

**Review by aelovikov-intel — APPROVED** [review]

File: SYCL/UserDefinedReductions/user_defined_reductions.cpp [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+
+      size_t temp_memory_size = workgroup_size * sizeof(InputT);
+      auto scratch = sycl::local_accessor<std::byte, 1>(temp_memory_size, cgh);
+      cgh.parallel_for(
+          nd_range<1>(workgroup_size, workgroup_size), [=](nd_item<1> it) {
+            // Create a handle that associates the group with an allocation it
+            // can use
+            auto handle =
+                sycl::ext::oneapi::experimental::group_with_scratchpad(
+                    it.get_group(), sycl::span(&scratch[0], temp_memory_size));
```
Comment (aelovikov-intel) [inline SYCL/UserDefinedReductions/user_defined_reductions.cpp:121]: I'm still not sure if it has non-pointer iterators underneath. Are you sure that scenario is tested now?

File: SYCL/UserDefinedReductions/user_defined_reductions.cpp [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+
+            InputT *first = in.get_pointer();
+            InputT *last = first + N;
```
Comment (aelovikov-intel) [inline SYCL/UserDefinedReductions/user_defined_reductions.cpp:124]: Non-pointer iterators seem to be untested still.

**Comment by pvchupin** [comment]
@dm-vodopyanov, new test fails on CUDA and HIP . Does it make sense disable it there?
